x/appchain/coordinator/types/msg.go

```diff
@@ -0,0 +1,86 @@
+package types
+
+import (
+	fmt "fmt"
+	"strings"
+
+	sdkerrors "cosmossdk.io/errors"
+	assetstypes "github.com/ExocoreNetwork/exocore/x/assets/types"
+	epochstypes "github.com/ExocoreNetwork/exocore/x/epochs/types"
+	"github.com/cometbft/cometbft/libs/json"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+const TypeRegisterSubscriberChain = "register_subscriber_chain"
+
+var _ sdk.Msg = &RegisterSubscriberChainRequest{}
+
+func (msg *RegisterSubscriberChainRequest) Route() string {
+	return RouterKey
+}
+
+func (msg *RegisterSubscriberChainRequest) Type() string {
+	return TypeRegisterSubscriberChain
+}
+
+// GetSignBytes returns the raw bytes for a RegisterSubscriberChainRequest message that
+// the expected signer needs to sign.
+func (msg *RegisterSubscriberChainRequest) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+// ValidateBasic executes sanity validation on the provided data
+// MsgUpdateParams is used to update params, the validation will mostly be stateful which is done by service
+func (msg *RegisterSubscriberChainRequest) ValidateBasic() error {
+	if _, err := sdk.AccAddressFromBech32(msg.FromAddress); err != nil {
+		return sdkerrors.Wrap(err, "invalid from address")
+	}
+	// the chainID must be non-empty; there is no maximum length imposed
+	chainID := strings.TrimSpace(msg.ChainID)
+	if val := len(chainID); val == 0 {
+		return fmt.Errorf("invalid chain id %s %d", chainID, val)
+	}
+	if epochstypes.ValidateEpochIdentifierString(msg.EpochIdentifier) != nil {
+		return fmt.Errorf("invalid epoch identifier %s", msg.EpochIdentifier)
+	}
+	for _, assetID := range msg.AssetIDs {
+		if _, _, err := assetstypes.ValidateID(
+			assetID,
+			true, // the caller must make them lowercase
+			true, // TODO: we support only Ethereum assets for now.
+		); err != nil {
+			return fmt.Errorf("invalid asset id %s: %s", assetID, err)
+		}
+	}
+	if msg.MaxValidators == 0 {
+		return fmt.Errorf("invalid max validators %d", msg.MaxValidators)
+	}
+	// no need to validate min self delegation, it is a uint64
+	if err := msg.SubscriberParams.Validate(); err != nil {
+		return fmt.Errorf("invalid subscriber params: %s", err)
+	}
+	return nil
+}
+
+// GetSigners returns the expected signers for a MsgUpdateParams message
+func (msg *RegisterSubscriberChainRequest) GetSigners() []sdk.AccAddress {
+	addr, err := sdk.AccAddressFromBech32(msg.FromAddress)
+	if err != nil {
+		// same behavior as cosmos-sdk
+		panic(err)
+	}
+	return []sdk.AccAddress{addr}
+}
+
+// NewRegisterSubscriberChainRequest creates a new RegisterSubscriberChainRequest using
+// the provided creator and json value.
+func NewRegisterSubscriberChainRequest(creator, jsonValue string) *RegisterSubscriberChainRequest {
+	var r RegisterSubscriberChainRequest
+	if err := json.Unmarshal([]byte(jsonValue), &r); err != nil {
+		panic(fmt.Sprintf("invalid json %s", err))
+	}
+	// the creator is overwritten
+	r.FromAddress = creator
+	return &r
```

Comment:
r.ValidateBasic() before return   ?

Reply:
Not necessary, since the Cosmos-SDK does the call before broadcast.